xarray/core/dataarray.py

```diff
@@ -1329,9 +1334,13 @@ def expand_dims(self, dim: Union[None, Hashable, Sequence[Hashable],
         ds = self._to_temp_dataset().expand_dims(dim, axis)
         return self._from_temp_dataset(ds)
 
-    def set_index(self, indexes: Optional[Mapping[Hashable, Any]] = None,
-                  append: bool = False, inplace: Optional[bool] = None,
-                  **indexes_kwargs: Any) -> Optional['DataArray']:
+    def set_index(
+        self,
+        indexes: Mapping[Hashable, Union[Hashable, Sequence[Hashable]]] = None,
+        append: bool = False,
+        inplace: bool = None,
+        **indexes_kwargs: Union[Hashable, Sequence[Hashable]]
```

Comment:
I'm guessing this is fine because mypy passes - my original thought was whether this should be `Optional[Mapping]` given it was either `None` or a dict?

Reply:
As of mypy 0.701, ``x: MyType = None`` and ``x: Optional[MyType] = None`` are the same.

```python
from typing import Optional


def f1(x: int = None) -> None:
    pass

def f2(x: Optional[int] = None) -> None:
    pass

def f3(x: int = 1) -> None:
    pass

def f4(x: Optional[int] = 1) -> None:
    pass


f1()
f2()
f3()
f4()
f1(1)
f2(1)
f3(1)
f4(1)
f1(None)
f2(None)
f3(None)
f4(None)
```
mypy output:
```
test_mypy.py:27: error: Argument 1 to "f3" has incompatible type "None"; expected "int"
```

Reply:
@max-sixty I did a generalised cleanup of all already existing cases; I find it's much more readable now